src/ORM/Paginator.php

```diff
@@ -0,0 +1,455 @@
+     * This helps prevent users from ordering large result sets on un-indexed values.
+     *
+     * If you need to sort on associated columns or synthetic properties you will need to use a whitelist.
+     *
+     * Any columns listed in the sort whitelist will be implicitly trusted. You can use this to sort
+     * on synthetic columns, or columns added in custom find operations that may not exist in the schema.
+     *
+     * @param \Cake\Datasource\RepositoryInterface $object Repository object.
+     * @param array $options The pagination options being used for this request.
+     * @return array An array of options with sort + direction removed and replaced with order if possible.
+     */
+    public function validateSort(RepositoryInterface $object, array $options)
+    {
+        if (isset($options['sort'])) {
+            $direction = null;
+            if (isset($options['direction'])) {
+                $direction = strtolower($options['direction']);
+            }
+            if (!in_array($direction, ['asc', 'desc'])) {
+                $direction = 'asc';
+            }
+            $options['order'] = [$options['sort'] => $direction];
+        }
+        unset($options['sort'], $options['direction']);
+
+        if (empty($options['order'])) {
+            $options['order'] = [];
+        }
+        if (!is_array($options['order'])) {
+            return $options;
+        }
+
+        $inWhitelist = false;
+        if (isset($options['sortWhitelist'])) {
+            $field = key($options['order']);
+            $inWhitelist = in_array($field, $options['sortWhitelist'], true);
+            if (!$inWhitelist) {
+                $options['order'] = [];
+
+                return $options;
+            }
+        }
+
+        $options['order'] = $this->_prefix($object, $options['order'], $inWhitelist);
+
+        return $options;
+    }
+
+    /**
+     * Prefixes the field with the table alias if possible.
+     *
+     * @param \Cake\Datasource\RepositoryInterface $object Repository object.
+     * @param array $order Order array.
+     * @param bool $whitelisted Whether or not the field was whitelisted
+     * @return array Final order array.
+     */
+    protected function _prefix(RepositoryInterface $object, $order, $whitelisted = false)
+    {
+        $tableAlias = $object->alias();
+        $tableOrder = [];
+        foreach ($order as $key => $value) {
+            if (is_numeric($key)) {
+                $tableOrder[] = $value;
+                continue;
+            }
+            $field = $key;
+            $alias = $tableAlias;
+
+            if (strpos($key, '.') !== false) {
+                list($alias, $field) = explode('.', $key);
+            }
+            $correctAlias = ($tableAlias === $alias);
+
+            if ($correctAlias && $whitelisted) {
+                // Disambiguate fields in schema. As id is quite common.
```

Comment:
While I am personally a fan of comments like this, the core is void of them, isn't it?